how-to-trace-linux-system-calls-in-production-with-minimal-impact-on-performance.md

```diff
@@ -0,0 +1,116 @@
+
++ To analyze the stack information of calls that have a high latency:
+
+    ```shell
+    perf trace record --call-graph dwarf -p $PID -- sleep 10
+    ```
+
+    ![*Stack information of system calls with high latency*](media/stack-information-of-system-calls-with-high-latency.jpg)
+
++ To trace a group of tasks. For example, two BPF tools are running in the background. To see their system call information, you can add them to a `perf_event` cgroup and then execute `per trace`:
+
+    ```shell
+    mkdir /sys/fs/cgroup/perf_event/bpftools/
+    echo 22542 >> /sys/fs/cgroup/perf_event/bpftools/tasks
+    echo 20514 >> /sys/fs/cgroup/perf_event/bpftools/tasks
+    perf trace -G bpftools -a -- sleep 10
+    ```
+
+    ![*Trace a group of tasks*](media/trace-a-group-of-tasks.jpg)
+
+    Those are some of the most common uses of perf.  If you’d like to know more (especially about perf-trace), see the [Linux manual page](https://man7.org/linux/man-pages/man1/perf-trace.1.html). From the manual pages, you will learn that perf-trace can filter tasks based on PIDs or thread IDs (TIDs), but that it has no convenient support for containers and the Kubernetes (K8s) environments. Don’t worry. Next, we’ll discuss a tool that can easily trace system calls in containers and in K8s environments that uses cgroup v2.
+
+## Traceloop, a performance profiler for cgroup v2 and K8s
+
+Traceloop provides better support for tracing Linux system calls in the containers or K8s environments that use cgroup v2. You might be unfamiliar with traceloop but know BPF Compiler Collection (BCC) pretty well. (Its front-end is implemented using Python or C++.) In the IO Visor Project, BCC’s parent project, there is another project named gobpf that provides Golang bindings for the BCC framework. Based on gobpf, traceloop is developed for environments of containers and K8s. The following illustration shows the traceloop architecture:
+
+![*traceloop architecture*](media/traceloop-architecture.jpg)
+
+We can further simplify this illustration into the following key procedures. Note that these procedures are implementation details, not operations to perform:
+
+1. `bpf helper` gets the cgroup ID. Tasks are filtered based on the cgroup ID rather than on the PID and TID.
+2. Each cgroup ID corresponds to a [bpf tail call](https://ebpf.io/what-is-ebpf/#tail--function-calls) that can call and execute another eBPF program and replace the execution context. Syscall events are written through a bpf tail call to a perf ring buffer with the same cgroup ID.
+3. The user space reads the perf ring buffer based on this cgroup ID.
+
+> **Note:**
+>
+> Currently, you  can get the cgroup ID only by executing `bpf helper: bpf_get_current_cgroup_id`, and this ID is available only in cgroup v2. Therefore, before you use traceloop, make sure that [cgroup v2 is enabled](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#basic-operations) in your environment.
```

Comment:
```suggestion
> Currently, you can get the cgroup ID only by executing `bpf helper: bpf_get_current_cgroup_id`, and this ID is available only in cgroup v2. Therefore, before you use traceloop, make sure that [cgroup v2 is enabled](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#basic-operations) in your environment.
```

Reply:
If this note is for bullet item #3, you must indent it.

Reply:
No, the note is for the whole list.